File: spectra/test/cpp/check_surface.cpp
#include <functional>

#include <polyregion/spectra_api.hpp>

void check(int *xs, const int *ys, const int *map) {
  float fs[4]{};
  double ds[4]{};
  const auto unary = [](const int &x) { return x + 1; };
  const auto binary = [](const int &x, const int &y) { return x + y; };
  const auto predicate = [](const int &x) { return x > 0; };
  const auto compare = [](const int &x, const int &y) { return x < y; };
  const auto valueOp = [](const float &x, const float &y) { return x + y; };

  spectra::transform(ys, fs, 4, [](const int &x) { return float(x + 1); });
  spectra::transform_binary(ys, fs, ds, 4, [](const int &x, const float &y) { return double(x) + y; });
  (void)spectra::reduce(ys, 4, 0, binary);
  spectra::exclusive_scan(ys, xs, 4, 0, binary);
  spectra::transform_inclusive_scan(ys, ds, 4, [](const int &x) { return double(x); }, std::plus<>{});
  spectra::transform_exclusive_scan(ys, ds, 4, 0.0, [](const int &x) { return double(x); }, std::plus<>{});
  spectra::adjacent_difference(ys, xs, 4, [](const int &x, const int &y) { return x - y; });
  spectra::for_each(xs, 4, unary);
  spectra::generate(xs, 4, [] { return 1; });
  spectra::copy(ys, 4, xs);
  spectra::gather(map, 4, ys, 16, xs);
  spectra::scatter(ys, 4, map, xs, 16);
  (void)spectra::all_of(ys, 4, predicate);
  (void)spectra::count(ys, 4, 1);
  (void)spectra::inner_product(ys, fs, 4, 0.0, std::plus<>{}, [](const int &x, const float &y) { return double(x) * y; });
  (void)spectra::transform_reduce(ys, 4, 0.0, [](const int &x) { return double(x); }, std::plus<>{});
  (void)spectra::set_intersection(ys, 4, ys, 4, xs, 4, compare);
  spectra::sort_by_key(xs, fs, 4, compare);
  (void)spectra::reduce_by_key(ys, fs, xs, fs, 4, compare, valueOp);
}

File: spectra/test/scala/polyregion/spectra/check_surface.scala
package polyregion.spectra

object check_surface {
  def compile(api: SpectraApi, input: Array[Int], output: Array[Int]): Unit = {
    val floats  = new Array[Float](input.length)
    val doubles = new Array[Double](input.length)
    api.transform(input, floats, input.length, _.toFloat + 1)
    api.transform_binary(input, floats, doubles, input.length, _.toDouble + _)
    api.reduce(input, input.length, 0, _ + _)
    api.exclusive_scan(input, output, input.length, 0, _ + _)
    api.transform_inclusive_scan(input, doubles, input.length, _.toDouble, _ + _)
    api.transform_exclusive_scan(input, doubles, input.length, 0.0, _.toDouble, _ + _)
    api.adjacent_difference(input, output, input.length, _ - _)
    api.for_each(output, output.length, _ + 1)
    api.generate(output, output.length, () => 1)
    api.copy(input, input.length, output)
    api.all_of(input, input.length, _ > 0)
    api.count(input, input.length, 1)
    api.inner_product(input, floats, input.length, 0.0, _ + _, _.toDouble * _)
    api.transform_reduce(input, input.length, 0.0, _.toDouble + 1, _ + _)
    api.set_intersection(input, input.length, input, input.length, output, output.length, _ < _)
    api.sort_by_key(input, floats, input.length, _ < _)
    api.reduce_by_key(input, floats, output, floats, input.length, _ == _, _ + _)
  }
}
